packages/client/src/generation/TSClient/Payload.ts

```diff
@@ -38,11 +45,34 @@ export type ${getPayloadName(name)}<
   }
   private renderRelations(projection: Projection): string {
     const { type } = this
+
     // TODO: can be optimized, we're calling the filter two times
-    const relations = type.fields.filter((f) => f.outputType.location === 'outputObjectTypes')
-    if (relations.length === 0 && projection === Projection.include) {
-      return ''
+    let relations: DMMF.SchemaField[] = []
+
+    // For select, a "relation"" can be a model or a composite
+    if (projection === Projection.select) {
+      relations = type.fields.filter((f) => {
+        return f.outputType.location === 'outputObjectTypes'
+      })
+    }
+
+    // But for include, a "relation" can only be a model
+    if (projection === Projection.include) {
+      const nonCompositeRelations = type.fields.filter((f) => {
+        return (
+          f.outputType.location === 'outputObjectTypes' &&
+          typeof f.outputType.type === 'object' &&
+          !this.dmmf.typeMap[f.outputType.type.name]
+        )
+      })
+
+      if (nonCompositeRelations.length === 0) {
+        return ''
+      } else {
+        relations = nonCompositeRelations
+      }
```

Comment:
I'd suggest using `ts-pattern` and its exhaustiveness check for this purpose, but even the following could do:

```suggestion
    const self = this

    const relations: DMMF.SchemaField[] = (() => {
      switch (projection) {
        case Projection.select:
          // For select, a "relation" can be a model or a composite type
          return type.fields.filter((f) => f.outputType.location === 'outputObjectTypes')
  
        case Projection.include:
          // For include, a "relation" can only be a model
          return type.fields.filter((f) => (
              f.outputType.location === 'outputObjectTypes'
              && typeof f.outputType.type === 'object'
              && !self.dmmf.typeMap[f.outputType.type.name]
            )
          )
        default:
          // I'd rather use `ts-pattern`'s `.exhaustive()` compile-time check here
          throw new Error(`Unknown projection ${projection}`)
        }
      })()

    if (relations.length === 0 && projection === Projection.include) {
      return ''
    }
```

In general, I'd advice against having free `let` declarations around, as well as missing `else` statements in enumeration checks

Reply:
One more thing you can use to ensure exhaustiveness with if/switch `assertNever` from `@prisma/internals`. [Example in TS playground](https://www.typescriptlang.org/play?#code/LAKApgHgDg9gTgFwAQDMCuA7AxggljDJAQwGcSxEA5MANwoAoi4BzALiQ1ooBokK54AWTBkizMOxII4uDMwCU7TnThIA3qCRIEACwEB3DmEMBROALj1+QkSTFh5oAL6hQAEzBYANkzBIsBFKocGi4COwA5ACCUFBeYBFIAD5IEQDycERyCcmpAApgTBGuICT6YVg6SPQoIWHy6pr+pH7RsfERrE1aAEZwhQDWANxNWC2pGVnind1IfYMjIFoeKERoXuGzpORUXJa1oQi8EQCqGAMYMPqEB2ER8otOQA). Notice compile-time error in default branch.

Reply:
Right, that's a viable alternative in this scenario

Reply:
I ended up with a hybrid between both suggestions.